Entities/SLTerrain.cpp

```diff
@@ -1,1436 +1,474 @@
+
+	// TODO: OPTIMIZE THIS, IT'S A TIME HOG. MAYBE JSUT STAMP THE OUTLINE AND SAMPLE SOME RANDOM PARTICLES?
+	std::deque<MOPixel *> SLTerrain::EraseSilhouette(BITMAP *sprite, const Vector &pos, const Vector &pivot, const Matrix &rotation, float scale, bool makeMOPs, int skipMOP, int maxMOPs) {
+		RTEAssert(sprite, "Null BITMAP passed to SLTerrain::EraseSilhouette");
+
+		int maxWidth = static_cast<int>(static_cast<float>(sprite->w + std::abs(pivot.GetFloorIntX() - (sprite->w / 2))) * scale);
+		int maxHeight = static_cast<int>(static_cast<float>(sprite->h + std::abs(pivot.GetFloorIntY() - (sprite->h / 2))) * scale);
+		int maxDiameter = static_cast<int>(std::sqrt(static_cast<float>(maxWidth * maxWidth + maxHeight * maxHeight)) * 2.0F);
+		int skipCount = skipMOP;
+
+		BITMAP *tempBitmap = g_SceneMan.GetIntermediateBitmapForSettlingIntoTerrain(maxDiameter);
+		clear_bitmap(tempBitmap);
+		pivot_scaled_sprite(tempBitmap, sprite, tempBitmap->w / 2, tempBitmap->h / 2, pivot.GetFloorIntX(), pivot.GetFloorIntY(), ftofix(rotation.GetAllegroAngle()), ftofix(scale));
+
+		std::deque<MOPixel *> dislodgedMOPixels;
+
+		// Test intersection between color pixels of the test bitmap and non-air pixels of the terrain, then generate and collect MOPixels that represent the terrain overlap and clear the same pixels out of the terrain.
+		for (int testY = 0; testY < tempBitmap->h; ++testY) {
+			for (int testX = 0; testX < tempBitmap->w; ++testX) {
+				int terrX = pos.GetFloorIntX() - (tempBitmap->w / 2) + testX;
+				int terrY = pos.GetFloorIntY() - (tempBitmap->h / 2) + testY;
+
+				if (terrX < 0) {
+					if (m_WrapX) {
+						while (terrX < 0) {
+							terrX += m_MainBitmap->w;
+						}
+					} else {
+						continue;
+					}
+				}
+				if (terrY < 0) {
+					if (m_WrapY) {
+						while (terrY < 0) {
+							terrY += m_MainBitmap->h;
+						}
+					} else {
+						continue;
+					}
+				}
+				if (terrX >= m_MainBitmap->w) {
+					if (m_WrapX) {
+						terrX %= m_MainBitmap->w;
+					} else {
+						continue;
+					}
+				}
+				if (terrY >= m_MainBitmap->h) {
+					if (m_WrapY) {
+						terrY %= m_MainBitmap->h;
+					} else {
+						continue;
+					}
+				}
+				int matPixel = getpixel(m_MainBitmap, terrX, terrY);
+				int colorPixel = getpixel(m_FGColorLayer->GetBitmap(), terrX, terrY);
+
+				if (getpixel(tempBitmap, testX, testY) != ColorKeys::g_MaskColor) {
+					// Only add PixelMO if we're not due to skip any.
+					if (makeMOPs && matPixel != MaterialColorKeys::g_MaterialAir && colorPixel != ColorKeys::g_MaskColor && ++skipCount > skipMOP && dislodgedMOPixels.size() < maxMOPs) {
+						skipCount = 0;
+						const Material *sceneMat = g_SceneMan.GetMaterialFromID(matPixel);
+						const Material *spawnMat = sceneMat->GetSpawnMaterial() ? g_SceneMan.GetMaterialFromID(sceneMat->GetSpawnMaterial()) : sceneMat;
+						// Create the MOPixel based off the Terrain data.
```

Comment:
This comment and the one below it aren't really needed. These parts of the code are fairly clear.